### PR TITLE
Fix chart of P_SEX

### PR DESCRIPTION
**Pull request by gn385x**

### CODE REVIEW

**Review by iamMoid — CHANGES_REQUESTED** [review]
Thanks for working on this. Let us use an `altair repeat` to create bar charts for each feature. We can use transform filter to create two sets of repeated charts for: Fatality=False and Fatality=True. I have the code here for you which you can use to update the script:

```
features = list(set(train_df.columns.values) - set(['FATALITY']))
features
``` 

```
Chart_False = alt.Chart(
    data=train_df
).mark_bar(
).encode(
    x=alt.X(alt.repeat('row'),
            type='quantitative',
            bin=alt.Bin(maxbins=40)),
    y='count()',
    color=alt.Color('FATALITY',
                    scale=alt.Scale(range=['green']),
                    legend=None)
).properties(
    width=150,
    height=100,
).repeat(
    row=features,
    columns=1,
    title="FATALITY = 'False'"
).resolve_scale(
    y='independent'
).transform_filter(
    alt.FieldOneOfPredicate(field='FATALITY',
                            oneOf=[False])
)

Chart_False
```

```
Chart_True = alt.Chart(
    data=train_df
).mark_bar(
).encode(
    x=alt.X(alt.repeat('row'),
            type='quantitative',
            bin=alt.Bin(maxbins=40)),
    y='count()',
    color=alt.Color('FATALITY',
                    scale=alt.Scale(range=['maroon']),
                    legend=None)
).properties(
    width=150,
    height=100
).repeat(
    row=features,
    columns=1,
    title="FATALITY = 'True'"
).resolve_scale(
    y='independent'
).transform_filter(
    alt.FieldOneOfPredicate(field='FATALITY',
                            oneOf=[True])
)

Chart_True
```

File: src/eda.py [hunk 1]
```diff
@@ -88,4 +88,7 @@ def create_and_save_chart(df, col, save_path):
     )
 
     for feature in train_df.columns:
-        create_and_save_chart(train_df, feature, save_path)
+        if feature == 'index' or feature == 'FATALITY':
```
Comment (danfke) [inline src/eda.py]: Shouldn't need feature=="index" because that feature has been set as the index in line 87

File: src/eda.py [hunk 2]
```diff
@@ -44,19 +44,19 @@ def create_and_save_chart(df, col, save_path):
 
     Examples
     --------
-    >>> create_chart('C_WTHR', 'Weather condition')
+    >>> create_chart(train_df, 'C_WTHR', results)
     """
 
-    if col == "P_AGE":
+    if col == "P_SEX":
```
Comment (danfke) [inline src/eda.py]: Will P_AGE work properly if you replace the if statement with P_SEX?

**Comment by gn385x** [comment]
> Thanks for working on this. Let us use an `altair repeat` to create bar charts for each feature. We can use transform filter to create two sets of repeated charts for: Fatality=False and Fatality=True. I have the code here for you which you can use to update the script:
> 
> ```
> features = list(set(train_df.columns.values) - set(['FATALITY']))
> features
> ```
> 
> ```
> Chart_False = alt.Chart(
>     data=train_df
> ).mark_bar(
> ).encode(
>     x=alt.X(alt.repeat('row'),
>             type='quantitative',
>             bin=alt.Bin(maxbins=40)),
>     y='count()',
>     color=alt.Color('FATALITY',
>                     scale=alt.Scale(range=['green']),
>                     legend=None)
> ).properties(
>     width=150,
>     height=100,
> ).repeat(
>     row=features,
>     columns=1,
>     title="FATALITY = 'False'"
> ).resolve_scale(
>     y='independent'
> ).transform_filter(
>     alt.FieldOneOfPredicate(field='FATALITY',
>                             oneOf=[False])
> )
> 
> Chart_False
> ```
> 
> ```
> Chart_True = alt.Chart(
>     data=train_df
> ).mark_bar(
> ).encode(
>     x=alt.X(alt.repeat('row'),
>             type='quantitative',
>             bin=alt.Bin(maxbins=40)),
>     y='count()',
>     color=alt.Color('FATALITY',
>                     scale=alt.Scale(range=['maroon']),
>                     legend=None)
> ).properties(
>     width=150,
>     height=100
> ).repeat(
>     row=features,
>     columns=1,
>     title="FATALITY = 'True'"
> ).resolve_scale(
>     y='independent'
> ).transform_filter(
>     alt.FieldOneOfPredicate(field='FATALITY',
>                             oneOf=[True])
> )
> 
> Chart_True
> ```

Thanks for the idea. I have incorporated it in the idea!

File: src/eda.py [hunk 3]
```diff
@@ -86,6 +30,51 @@ def create_and_save_chart(df, col, save_path):
...
+        .properties(width=150, height=150)
+        .repeat(
+            row=features,
+            columns=1,
+            title="FATALITY = True")
+        .resolve_scale(y="independent")
+        .transform_filter(alt.FieldOneOfPredicate(field="FATALITY", oneOf=[True]))
+    )
+    
+    Chart_True.save(f"{save_path}/Distribution_of_fatality.png")
```
Comment (danfke) [inline src/eda.py]: Get rid of the slash after save_path to keep it consistent with other scripts and the README

**Review by danfke — COMMENTED** [review]
Looks good to me, I don't really know how the code works but given that you and Moid have gone through it I trust it does

File: src/eda.py [hunk 4]
```diff
@@ -86,6 +30,51 @@ def create_and_save_chart(df, col, save_path):
...
+        .properties(width=150, height=150)
+        .repeat(
+            row=features,
+            columns=1,
+            title="FATALITY = False")
+        .resolve_scale(y="independent")
+        .transform_filter(alt.FieldOneOfPredicate(field="FATALITY", oneOf=[False]))
+    )
+    
+    Chart_False.save(f"{save_path}/Distribution_of_no_fatality.png")
```
Comment (danfke) [inline src/eda.py]: Get rid of slash after save path to keep consistent with other scripts and README

File: src/eda.py (hunk 3 above)
Comment (gn385x) [reply src/eda.py]: > Looks good to me, I don't really know how the code works but given that you and Moid have gone through it I trust it does

Distribution_of_fatality

Chart_True looks like this Daniel. We have a graph similar for Chart_False.

File: src/eda.py (hunk 4 above)
Comment (gn385x) [reply src/eda.py]: done

File: src/eda.py (hunk 1 above)
Comment (gn385x) [reply src/eda.py]: not applicable to new code

File: src/eda.py (hunk 2 above)
Comment (gn385x) [reply src/eda.py]: not applicable to new code

**Comment by SiqiTao** [comment]
Thanks! @gn385x  Pretty sure the current code works well,  but I am still a bit confused about the code structure --- are we supposed to write a separate main function (`def main()`) as demonstrated in lectures instead of putting the whole `main()` chunk inside an if statement?

**Comment by gn385x** [comment]
> Thanks! @gn385x Pretty sure the current code works well, but I am still a bit confused about the code structure --- are we supposed to write a separate main function (`def main()`) as demonstrated in lectures instead of putting the whole `main()` chunk inside an if statement?

main() is just the name of the function. like how we have create_and_save_chart() function before. that main() is different from the __main__ in "if __name__ == "__main__":". we don't have to create a function if we don't need to. (also got this confirmed by TA during lab).

File: src/eda.py (hunk 3 above)
Comment (danfke) [reply src/eda.py]: Why is there overlap in the bars?

**Comment by SiqiTao** [comment]
> 

Thanks for the explanation! The code looks fine to me now.

**Comment by SiqiTao** [comment]
> > Thanks! @gn385x Pretty sure the current code works well, but I am still a bit confused about the code structure --- are we supposed to write a separate main function (`def main()`) as demonstrated in lectures instead of putting the whole `main()` chunk inside an if statement?
> 
> main() is just the name of the function. like how we have create_and_save_chart() function before. that main() is different from the **main** in "if **name** == "**main**":". we don't have to create a function if we don't need to. (also got this confirmed by TA during lab).

Thanks for the explanation! The code looks fine to me now.

File: src/eda.py (hunk 3 above)
Comment (iamMoid) [reply src/eda.py]: Guess it is the bin size at play. Can you look into it @gn385x?

Comment (gn385x) [reply src/eda.py]: it's the bin size. fixed!

**Review by iamMoid — APPROVED** [review]
Looks good. Thanks for updating!

**Review by SiqiTao — APPROVED** [review]

**Pull request merged**